packages/vaadin-virtual-list/src/vaadin-virtual-list.js

```diff
@@ -0,0 +1,141 @@
+ * ```
+ *
+ * ```js
+ * const list = document.querySelector('vaadin-virtual-list');
+ * list.items = items; // An array of data items
+ * list.renderer = (root, list, {item, index}) => {
+ *   root.textContent = `#${index}: ${item.name}`
+ * }
+ * ```
+ *
+ * See [Virtual List](https://vaadin.com/docs/latest/ds/components/virtual-list) documentation.
+ *
+ * @extends HTMLElement
+ * @mixes ElementMixin
+ * @mixes ThemableMixin
+ */
+class VirtualListElement extends ElementMixin(ThemableMixin(PolymerElement)) {
+  static get template() {
+    return html`
+      <style>
+        :host {
+          display: block;
+          height: 200px;
+          overflow: auto;
+        }
+
+        :host([hidden]) {
+          display: none !important;
+        }
+      </style>
+
+      <div id="items">
+        <slot></slot>
+      </div>
+    `;
+  }
+
+  static get is() {
+    return 'vaadin-virtual-list';
+  }
+
+  static get version() {
+    return '21.0.0-alpha2';
+  }
+
+  static get properties() {
+    return {
+      /**
+       * An array containing items determining how many instances to render.
+       * @type {Array<!VirtualListItem> | undefined}
+       */
+      items: { type: Array },
+
+      /**
+       * Custom function for rendering the content of every item.
+       * Receives three arguments:
+       *
+       * - `root` The render target element representing one item at a time.
+       * - `virtualList` The reference to the `<vaadin-virtual-list>` element.
+       * - `model` The object with the properties related with the rendered
+       *   item, contains:
+       *   - `model.index` The index of the rendered item.
+       *   - `model.item` The item.
+       * @type {VirtualListRenderer | undefined}
+       */
+      renderer: Function
+    };
+  }
+
+  static get observers() {
+    return ['__itemsOrRendererChanged(items, renderer)'];
+  }
+
+  /** @protected */
+  ready() {
+    super.ready();
+
+    this.__virtualizer = new Virtualizer({
+      createElements: this.__createElements,
+      updateElement: this.__updateElement.bind(this),
+      elementsContainer: this,
+      scrollTarget: this,
+      scrollContainer: this.shadowRoot.querySelector('#items')
+    });
+
+    if (window.Vaadin && window.Vaadin.templateRendererCallback) {
+      window.Vaadin.templateRendererCallback(this);
+    }
+  }
+
+  /**
+   * Scroll to a specific index in the virtual list.
+   *
+   * @param {number} index Index to scroll to
+   */
+  scrollToIndex(index) {
+    this.__virtualizer.scrollToIndex(index);
+  }
+
+  /** @private */
+  __createElements(count) {
+    return [...Array(count)].map(() => document.createElement('div'));
+  }
+
+  /** @private */
+  __updateElement(el, index) {
+    if (this.renderer) {
+      this.renderer(el, this, { item: this.items[index], index });
+    }
+  }
+
+  /** @private */
+  __itemsOrRendererChanged(items = [], renderer) {
+    if (renderer) {
+      if (items.length === this.__virtualizer.size) {
```

Comment:
Let's add a check for `this.__virtualizer` here as it is created in `ready` and the observer could run earlier.

See #379 where a similar issue for `vaadin-notification` was fixed.

Reply:
Done